common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcherTest.kt

```diff
@@ -62,11 +61,9 @@ class PurchasedProductsFetcherTest {
             productIds = listOf("product1", "product2"),
             purchaseTime = testDate.time
         )
-
         mockActivePurchases(activePurchase)
         mockAllPurchases(purchaseRecord)
-
-        var receivedListOfPurchasedProducts: List<PurchasedProduct>? = null
+        var receivedListOfPurchasedProducts: List<PurchasedProduct> = emptyList()
```

Comment:
There is an advantage of keeping it null originally. That is, we know that the success callback was called. Otherwise, the test would still fail, but we are not sure whether the callback was not called or was called with an empty list with further investigation. Same for the other similar changes in this PR.

We can potentially change the `!!` with `?` in most assertions since that would also fail the assertion. Thoughts?

Reply:
> That is, we know that the success callback was called. Otherwise, the test would still fail, but we are not sure whether the callback was not called or was called with an empty list with further investigation.

Not really. If the success callback is not called, the `onError` one will, and will fail the test right away https://github.com/RevenueCat/purchases-android/blob/2f746b6f87431979a6cba8eb8738ee1865f63f3e/common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcherTest.kt#L38

Note that we're stubbing the `onSuccess` callback https://github.com/RevenueCat/purchases-android/blob/2f746b6f87431979a6cba8eb8738ee1865f63f3e/common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcherTest.kt#L269 and we're also testing for when there are no purchases https://github.com/RevenueCat/purchases-android/blob/2f746b6f87431979a6cba8eb8738ee1865f63f3e/common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcherTest.kt#L72-L96 which is returning `onSuccess` with an empty list. `queryActiveProducts` can't return `null` https://github.com/RevenueCat/purchases-android/blob/2f746b6f87431979a6cba8eb8738ee1865f63f3e/common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcher.kt#L23

Question here is if that's expected. If so, we're already testing such scenario (`null` `receivedListOfPurchasedProducts`) and it should be safe to set  `receivedListOfPurchasedProducts` as non-null for the rest. Does that make sense or am I missing something?

Reply:
> It could happen that no callbacks are called in some situation (it has happened), which is what making it nullable and asserting it's not null would check.

If that's the case, we should have a separate test that only checks for that, in isolation I mean. I don't see any value of adding an extra assertion on every other test, especially if it's already tested and we know for sure it won't fail, quite the opposite, in my opinion, we're adding complexity and unnecessary time to the test suite.

---

On a related topic, generally is preferable having only 1 assertion per test 👀 https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=57

The problem of having multiple assertions are:
- It's annoying to have to look at the stacktrace to know what's failing (which assertion failed?)
- You don't have information about the remaining assertions (because when an assertion fails it breaks the execution of the test)

That's why is preferable to have only 1 assertion per test and one test per scenario. Basically, if one of them fails you'll know at a glance what in the code might have caused that failure and in which part (class, method, etc.).

I rarely write tests with multiple assertions (at least I try 😅), only if they're tearing apart the result object to verify everything e.g. if there's no `equals` (which is another problem although could happen if it's coming from a 3rd party library object) or if the assertions are atomic to the object.

This is a technique that I learnt from the book [Working Effectively With Unit Tests](https://leanpub.com/wewut). I really like the techniques described on it in favor of DAMP (Descriptive And Maintainable Procedures). Mainly because they increase maintainability and readability of tests. But that's my opinion and we can discuss further 😉 

BTW, I highly recommend you reading that book (if you haven't already). It's 🔝 💯

---

Rebased and re-pushed including the changes we're discussing. Please, let me know what you think when you have a chance 👍

Reply:
> If that's the case, we should have a separate test that only checks for that, in isolation I mean. I don't see any value of adding an extra assertion on every other test, especially if it's already tested and we know for sure it won't fail, quite the opposite, in my opinion, we're adding complexity and unnecessary time to the test suite.

True, I agree that tests should ideally have only one assertion to improve legibility. I see you already added tests to check for that specifically 👍. 

Something to consider, however, is that, if there are different situations where callbacks get called we should test that for every situation. For example, if there are different places in the tested code where the `onError` callback gets called, we should have this test for each of those (or have the callback only called once, which depending on the situation, might require some refactors). Otherwise we might miss cases. 

Having said that, there is definitely room for improvement in our tests, and I think this is a good step!

Reply:
> Something to consider, however, is that, if there are different situations where callbacks get called we should test that for every situation. For example, if there are different places in the tested code where the onError callback gets called, we should have this test for each of those (or have the callback only called once, which depending on the situation, might require some refactors). Otherwise we might miss cases.

Yeah totally agree! There are cases in which you can't have only one assertion per test because you wouldn't cover all scenarios. Don't get me wrong, this can't be applied blindly, just one more thing to take into account when writing tests 👍 

> Having said that, there is definitely room for improvement in our tests, and I think this is a good step!

Saw that you accepted the contribution, thanks much! Happy to keep helping 😊